lib/Frontend/Frontend.cpp

```diff
@@ -69,6 +69,51 @@ void CompilerInstance::setPrimarySourceFile(SourceFile *SF) {
   PrimarySourceFile->setReferencedNameTracker(NameTracker);
 }
 
+Optional<std::string>
+CompilerInstance::getPCHFilename() const {
+  auto &Importer =
+    *static_cast<ClangImporter *>(Context->getClangModuleLoader());
+  const auto &BridgingHeader =
+    Invocation.getClangImporterOptions().BridgingHeader;
+  const auto &PCHOutputDir =
+    Invocation.getClangImporterOptions().PrecompiledHeaderOutputDir;
+  if (BridgingHeader.empty() || PCHOutputDir.empty()) {
+    return None;
+  }
+
+  SmallString<256> PCHBasename { llvm::sys::path::filename(BridgingHeader) };
+  llvm::sys::path::replace_extension(PCHBasename, "");
+  PCHBasename.append("-swift_");
+  PCHBasename.append(Invocation.getPCHHash());
+  PCHBasename.append("-clang_");
+  PCHBasename.append(Importer.getClangModuleHash());
+  llvm::sys::path::replace_extension(PCHBasename, ".pch");
+  SmallString<256> PCHFilename { PCHOutputDir };
+  llvm::sys::path::append(PCHFilename, PCHBasename);
+  return PCHFilename.str().str();
+}
+
+void CompilerInstance::getOrCreatePCH() {
+  auto &Importer = *static_cast<ClangImporter *>(Context->getClangModuleLoader());
+  auto PCHFilename = getPCHFilename();
+  if (!PCHFilename.hasValue()) {
+    return;
+  }
+  if (!Importer.canReadPCH(PCHFilename.getValue())) {
+    auto FailedToEmit =
+    Importer.emitBridgingPCH(
+      Invocation.getClangImporterOptions().BridgingHeader,
+      PCHFilename.getValue());
+    if (FailedToEmit) {
+      return;
+    }
+  }
+
+  Invocation.getFrontendOptions().ImplicitObjCHeaderPath =
+    PCHFilename.getValue();
+  Invocation.getClangImporterOptions().BridgingHeader = PCHFilename.getValue();
```

Comment:
Rather than reading your mind, are there any other concerns you can share before I make another pass at this? Are you happy with how things are going? I would very much like to not waste any more time contributing nonsense.